### PR TITLE
Adding AbstractJobSpecScheduler

### PR DESCRIPTION
**Pull request by chavdar**
- Change JobSpecSchedulerListener.onJobTriggered(JobSpecSchedule jobSchedule) to onJobTriggered(JobSpec jobSpec)
- Add AbstractJobSpecScheduler to do some of the boiler-plate processing in schedulers

@pcadabam @ibuenros can you review?

### CODE REVIEW

**Comment by chavdar** [comment]
There are a lot of trivial changes in files due to the change in API. The meat of the PR is in AbstractJobSpecScheduler.

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+
+import gobblin.runtime.api.JobSpec;
+import gobblin.runtime.api.JobSpecSchedule;
+import gobblin.runtime.api.JobSpecScheduler;
+import gobblin.runtime.api.JobSpecSchedulerListener;
+
+/**
+ * A base implementation of {@link JobSpecScheduler} that keeps track
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java]: ++ that keeps track of scheduled jobs?

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ */
+public abstract class AbstractJobSpecScheduler implements JobSpecScheduler {
+  protected final Map<URI, JobSpecSchedule> _schedules = new HashMap<>();
+  private final Logger _log;
+  private final JobSpecSchedulerListeners _callbacksDispatcher;
+
+  public AbstractJobSpecScheduler(Optional<Logger> log) {
+    _log = log.isPresent() ? log.get() : LoggerFactory.getLogger(getClass());
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java]: ```log.or(LoggerFactory.getLogger(getClass()))```

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/ImmediateJobSpecScheduler.java [hunk 3]
```diff
@@ -11,101 +11,49 @@
...
-      _callbacksDispatcher.onJobTriggered(jobSchedule);
-      this.jobSchedule.getJobRunnable().run();
-    }
-  }
-
   @Override
-  public void registerWeakJobSpecSchedulerListener(JobSpecSchedulerListener listener) {
-    _callbacksDispatcher.registerWeakJobSpecSchedulerListener(listener);
+  protected void doUnschedule(JobSpecSchedule existingSchedule) {
+    // nothing to do
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/ImmediateJobSpecScheduler.java]: don't you need to remove the schedule? Is this a run immediately and run once scheduler?

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java (hunk 1 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java]: Fixed.

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java (hunk 2 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/AbstractJobSpecScheduler.java]: Fixed.

File: gobblin-runtime/src/main/java/gobblin/runtime/scheduler/ImmediateJobSpecScheduler.java (hunk 3 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/scheduler/ImmediateJobSpecScheduler.java]: The removal of the schedule happens inAbstractJobSpecScheduler . 

Yes, it is a immediately once scheduler.

**Comment by pcadabam** [comment]
LGTM

**Pull request merged**